### ISSUE TITLE
No consistant way to obtain hand and finger positions when controllers are used

### ISSUE THREAD

**Issue by BastiaanOlij** [KhronosGroup/OpenXR-Docs#120]
Currently only SteamVR (and possibly WMR) will provide an inferred hand skeleton through the hand tracking API when controllers are used that will allow a game developer to correctly position the players hand and fingers in relation to the controller.

Other platforms require the game developer to implement this themselves when controllers are used but there are a number of problems with this approach:
1) OpenXR does not define a specification for where the hand should be placed in relation to the controller, the grip pose does not seem usable for this.
2) OpenXR does not define a specification for retrieving controller dimensions that allow the game developer to properly place fingers, for instance differences in sizes between Touch, Quest 1 and Quest 2 controllers leads to issues in properly placing fingers on buttons.
3) OpenXR does not define a specification for obtaining raw input of the controller, this leads to issues for game engine builders who wish to implement this logic but where the game developer may define an action map that does not not provide access to the needed inputs.

Having the XR runtime provide a skeleton that allows the game developer to properly place the hands regardless of whether physical hand tracking is used or if the hand and finger positions are inferred from controller tracking and input, would ensure portability between the different platforms and solve the need for game developers to know much more about the XR system used than is currently provided and for which previous decisions exist that access to the needed data is unwanted.

There are however a number of shortcomings in the specification that need to be addressed:
1) Obviously the OpenXR specification needs to require inferred hand tracking from controller positions is uniformly supported.
2) While OpenXR defines a specification for the required bones and their relations, differences in interpretation between the different platforms prevents portability between the platforms.  A properly skinned hand for SteamVR, will not work properly for Oculus nor WMR. More clarity is needed in the specification to ensure the different XR runtimes return a hand skeleton that is usable with the same 3D asset
3) OpenXR does not define a specification that allows the user to know whether physical hand tracking is used or whether the hand tracking is inferred. This can be important if hand tracking is used for direct interactions between the hand and game world. Inferred finger positions can lead to sudden positional changes when the user lets go of a button on the controller so the game developer needs to be able to react accordingly.
4) OpenXR does not define a specification that allows the game (engine) developer to deal with the difference in hand sizes between players that is pulled through when physical hand tracking is used. Multiple options are possible from ignoring the players hand size and re-targeting to a predefined uniform hand, to offering an option for the XR runtime to provide a properly skinned hand asset that matches the players hand size.

It would be great if the OpenXR specification could deal with these issues so true portability between platforms is possible.

**Comment by rpavlik** [comment]
I'll bring these to the attention of those working on hand stuff. FYI 1.0.24 did include XR_EXT_palm_pose which helps slightly with some "put a hand model where it would be physically" but not most of your questions.

**Comment by BastiaanOlij** [comment]
Just capturing some feedback on this after discussing this with various people.

**Issues 1 + 3**
There seems to be a willingness to go down this path and ensure that bone information becomes available for hands regardless of whether the data is inferred from controller input or coming from full visual tracking of hands / gloves / other full hand tracking solutions. This would result in an application being able to rely on the bone data to visualize the hand provided the hand tracking extension is supported by the runtime.

There also seems to be agreement on having an ability to know whether bone data is inferred or real, the method by which is still being discussed.

**Issues 2 + 4**
Visual hand tracking by its nature has the result of providing bone data that is real, that means the dimensions of the hand will conform to the users real hand and thus vary in size.

Only inferred hand tracking has the ability to work from a fixed size skeleton.

The consensus seems to be that these are ultimately issues to be resolved on the XR client side, not by the XR runtime, and I tent to agree with this position.

I would like to suggest that as part of the render models extension we should evaluate whether it is possible that the XR runtime provides a render model for the hands.

**Comment by ChristophHaag** [comment]
Yes, you should be able to plug any space into here https://github.com/GodotVR/godot_openxr/blob/a3fd9c75febab26b421177ca27030f9ec54f467a/src/openxr/extensions/xr_ext_hand_tracking_extension_wrapper.cpp#L290, including the /user/hand/*/grip/pose action space or palm pose (if available).

You can't get the rest of the joint poses relative to the palm joint pose though, only relative to a XrSpace.
